### ISSUE TITLE
jank start button polling

### ISSUE THREAD

**Issue by ToruTheRedFox** [xan1242/NFS-XtendedInput#8]
it doesn't pause the game half the time you press it

**Comment by xan1242** [comment]
I need more info on this. I'm testing with an Xbox One controller and it works fine.

Pause action polls on key down action only, not on key up.

**Comment by ToruTheRedFox** [comment]
with a DS3 emulating a 360 controller it only sometimes works

**Comment by AeroWidescreen** [comment]
I initially had the same problem, for what it's worth. But this was back when I was using one of the early test builds you gave me. I can't reproduce the problem anymore. 

![Pause](https://i.imgur.com/1iu3tA3.gif)

**Comment by ToruTheRedFox** [comment]
I'm getting low framerates because of reshade so it might be an issue related to framerate?


try bogging the game down with a ridiculously overcomplicated reshade shader (idk add a for loop that does nothing for 32 iterations per pixel or smth idk lmao)

I'll do more tests and report back

**Comment by ToruTheRedFox** [comment]
here's my configuration for reference: 
```
[Input]
KeyboardReadingMode = 0 // 0 = async unbuffered, 1 = buffered
ConfineMouse = 0 // confine mouse within the game window
UseWin32Cursor = 0 // use Win32/OS to render the mouse (recommended for performance benefits)
UseCustomCursor = 0 // use the nfs_cursor.cur file as the cursor style
DeadzonePercentLS = 0.24 // set the percentage of deadzone for each stick here (default 24% = 0.24)
DeadzonePercentRS = 0.24
DeadzonePercentLS_P2 = 0.24
DeadzonePercentRS_P2 = 0.24
DeadzonePercent_Shifting = 0.75 // deadzone for shifting digital activation
DeadzonePercent_AnalogStickDigital = 0.50 // deadzone for analog sticks digital activation
DeadzonePercent_AnalogTriggerDigital = 0.12 // same for triggers

[Icons]
ControllerIconMode = 1 // 0 = Xbox One, 1 = PlayStation 4
FirstControlDevice = 1 // 0 = Keyboard, 1 = Controller -- for setting the first state

[Events]
FRONTENDACTION_UP = XINPUT_GAMEPAD_DPAD_UP
FRONTENDACTION_DOWN = XINPUT_GAMEPAD_DPAD_DOWN
FRONTENDACTION_LEFT = XINPUT_GAMEPAD_DPAD_LEFT
FRONTENDACTION_RIGHT = XINPUT_GAMEPAD_DPAD_RIGHT
FRONTENDACTION_UP_ALT = XINPUT_GAMEPAD_LS_UP
FRONTENDACTION_DOWN_ALT = XINPUT_GAMEPAD_LS_DOWN
FRONTENDACTION_LEFT_ALT = XINPUT_GAMEPAD_LS_LEFT
FRONTENDACTION_RIGHT_ALT = XINPUT_GAMEPAD_LS_RIGHT
FRONTENDACTION_ACCEPT = XINPUT_GAMEPAD_A
FRONTENDACTION_CANCEL = XINPUT_GAMEPAD_B
FRONTENDACTION_RUP = XINPUT_GAMEPAD_RS_UP
FRONTENDACTION_RDOWN = XINPUT_GAMEPAD_RS_DOWN
FRONTENDACTION_RLEFT = XINPUT_GAMEPAD_RS_LEFT
FRONTENDACTION_RRIGHT = XINPUT_GAMEPAD_RS_RIGHT
FRONTENDACTION_BUTTON0 = XINPUT_GAMEPAD_RIGHT_SHOULDER
FRONTENDACTION_BUTTON1 = XINPUT_GAMEPAD_LEFT_SHOULDER
FRONTENDACTION_BUTTON2 = XINPUT_GAMEPAD_LEFT_THUMB
FRONTENDACTION_BUTTON3 = XINPUT_GAMEPAD_RIGHT_THUMB
FRONTENDACTION_BUTTON4 = XINPUT_GAMEPAD_X
FRONTENDACTION_BUTTON5 = XINPUT_GAMEPAD_Y
FRONTENDACTION_LTRIGGER = XINPUT_GAMEPAD_LT
FRONTENDACTION_RTRIGGER = XINPUT_GAMEPAD_RT
FRONTENDACTION_START = XINPUT_GAMEPAD_START
// In-Game controls
GAMEACTION_GAS = XINPUT_GAMEPAD_RT
GAMEACTION_BRAKE = XINPUT_GAMEPAD_LT
GAMEACTION_STEERLEFT = XINPUT_GAMEPAD_LS_LEFT
GAMEACTION_STEERRIGHT = XINPUT_GAMEPAD_LS_RIGHT
GAMEACTION_TURNLEFT = XINPUT_GAMEPAD_LS_LEFT
GAMEACTION_TURNRIGHT = XINPUT_GAMEPAD_LS_RIGHT
GAMEACTION_HANDBRAKE = XINPUT_GAMEPAD_A
GAMEACTION_NOS = XINPUT_GAMEPAD_B
GAMEACTION_GAMEBREAKER = XINPUT_GAMEPAD_LEFT_SHOULDER
GAMEACTION_SHIFTUP = XINPUT_GAMEPAD_Y
GAMEACTION_SHIFTDOWN = XINPUT_GAMEPAD_X
GAMEACTION_RESET = XINPUT_GAMEPAD_BACK
HUDACTION_PAUSEREQUEST = XINPUT_GAMEPAD_START
HUDACTION_ENGAGE_EVENT = XINPUT_GAMEPAD_DPAD_UP
HUDACTION_PAD_LEFT = XINPUT_GAMEPAD_DPAD_LEFT
HUDACTION_PAD_DOWN = XINPUT_GAMEPAD_DPAD_DOWN
HUDACTION_PAD_RIGHT = XINPUT_GAMEPAD_DPAD_RIGHT
HUDACTION_SKIPNIS = XINPUT_GAMEPAD_A
HUDACTION_NEXTSONG = XINPUT_GAMEPAD_RIGHT_THUMB
CAMERAACTION_CHANGE = XINPUT_GAMEPAD_RS_DOWN
CAMERAACTION_LOOKBACK = XINPUT_GAMEPAD_RIGHT_SHOULDER
// Debug camera
CAMERAACTION_DEBUG = XINPUT_GAMEPAD_BACK
DEBUGACTION_DROPCAR = XINPUT_GAMEPAD_START
DEBUGACTION_MOVE_FORWARD = XINPUT_GAMEPAD_LS_UP
DEBUGACTION_MOVE_BACK = XINPUT_GAMEPAD_LS_DOWN
DEBUGACTION_MOVE_LEFT = XINPUT_GAMEPAD_LS_LEFT
DEBUGACTION_MOVE_RIGHT = XINPUT_GAMEPAD_LS_RIGHT
DEBUGACTION_MOVE_UP = XINPUT_GAMEPAD_RIGHT_SHOULDER
DEBUGACTION_MOVE_DOWN = XINPUT_GAMEPAD_LEFT_SHOULDER
DEBUGACTION_LOOK_UP = XINPUT_GAMEPAD_RS_DOWN
DEBUGACTION_LOOK_DOWN = XINPUT_GAMEPAD_RS_UP
DEBUGACTION_LOOK_LEFT = XINPUT_GAMEPAD_RS_LEFT
DEBUGACTION_LOOK_RIGHT = XINPUT_GAMEPAD_RS_RIGHT
DEBUGACTION_TURBO = XINPUT_GAMEPAD_LT
DEBUGACTION_SUPER_TURBO = XINPUT_GAMEPAD_RT
DEBUGACTION_LOOK_D_UP = XINPUT_GAMEPAD_DPAD_UP
DEBUGACTION_LOOK_D_DOWN = XINPUT_GAMEPAD_DPAD_DOWN
DEBUGACTION_LOOK_D_LEFT = XINPUT_GAMEPAD_DPAD_LEFT
DEBUGACTION_LOOK_D_RIGHT = XINPUT_GAMEPAD_DPAD_RIGHT
DEBUGACTION_MOVE_D_FORWARD = XINPUT_GAMEPAD_Y
DEBUGACTION_MOVE_D_BACK = XINPUT_GAMEPAD_A
DEBUGACTION_MOVE_D_LEFT = XINPUT_GAMEPAD_X
DEBUGACTION_MOVE_D_RIGHT = XINPUT_GAMEPAD_B
[EventsKB]
FRONTENDACTION_UP = VK_UP
FRONTENDACTION_DOWN = VK_DOWN
FRONTENDACTION_LEFT = VK_LEFT
FRONTENDACTION_RIGHT = VK_RIGHT
FRONTENDACTION_ACCEPT = VK_RETURN
FRONTENDACTION_CANCEL = VK_ESCAPE
FRONTENDACTION_RUP = W
FRONTENDACTION_RDOWN = S
FRONTENDACTION_RLEFT = A
FRONTENDACTION_RRIGHT = D
FRONTENDACTION_BUTTON0 = 3
FRONTENDACTION_BUTTON1 = 5
FRONTENDACTION_BUTTON2 = T
FRONTENDACTION_BUTTON3 = R
FRONTENDACTION_BUTTON4 = 2
FRONTENDACTION_BUTTON5 = 1
FRONTENDACTION_LTRIGGER = 9
FRONTENDACTION_RTRIGGER = 0
FRONTENDACTION_START = 4
// In-Game controls
GAMEACTION_GAS = VK_UP
GAMEACTION_BRAKE = VK_DOWN
GAMEACTION_STEERLEFT = VK_LEFT
GAMEACTION_STEERRIGHT = VK_RIGHT
GAMEACTION_TURNLEFT = VK_LEFT
GAMEACTION_TURNRIGHT = VK_RIGHT
GAMEACTION_HANDBRAKE = VK_SPACE
GAMEACTION_NOS = VK_MENU
GAMEACTION_GAMEBREAKER = X
GAMEACTION_SHIFTUP = VK_LSHIFT
GAMEACTION_SHIFTDOWN = VK_LCONTROL
GAMEACTION_RESET = R
HUDACTION_PAUSEREQUEST = VK_ESCAPE
HUDACTION_ENGAGE_EVENT = VK_RETURN
HUDACTION_PAD_LEFT = M
HUDACTION_PAD_RIGHT = VK_TAB
HUDACTION_PAD_DOWN = B
CAMERAACTION_CHANGE = C
CAMERAACTION_LOOKBACK = L
CAMERAACTION_PULLBACK = P
// Debug camera
CAMERAACTION_DEBUG = VK_SUBTRACT
DEBUGACTION_DROPCAR = 5
DEBUGACTION_MOVE_FORWARD = W
DEBUGACTION_MOVE_BACK = S
DEBUGACTION_MOVE_LEFT = A
DEBUGACTION_MOVE_RIGHT = D
DEBUGACTION_MOVE_UP = VK_SPACE
DEBUGACTION_MOVE_DOWN = VK_LCONTROL
DEBUGACTION_LOOK_UP = K
DEBUGACTION_LOOK_DOWN = I
DEBUGACTION_LOOK_LEFT = J
DEBUGACTION_LOOK_RIGHT = L
DEBUGACTION_TURBO = VK_LSHIFT
DEBUGACTION_SUPER_TURBO = F

```

**Comment by xan1242** [comment]
The device polling runs in the exact same place in the code where the original was (which is within the main thread). It is beefier than the original, yes, but really not by enough to exhibit this behavior.

These issues should (and probably will) happen without this plugin on low framerate in any case (and I can confirm this from running this game on a Pentium 1 MMX). It is standard behavior on low tick/frame rates to behave like this.

The only real "solution" that I can think of is to make this multi-threaded which would cause other issues, such as synchronization problems with the game's own queue.

**Comment by AeroWidescreen** [comment]
Hmm... well I got it to happen to again but it's very inconsistent. When I press the "start" button, _sometimes_ it doesn't register my input. I have to press it twice. But I have no reliable way to reproduce this, it just happens at random apparently.

**Comment by ToruTheRedFox** [comment]
it's really inconsistent and only happens with pausing
every other action (reset, shift gears, NOS, handbrake, etc.) works without issue

**Comment by AeroWidescreen** [comment]
I'm not sure how much this will help, but here's a GIF of it happening.

![PauseBug](https://i.imgur.com/mtwb5KV.gif)

As you can see, the first input doesn't do anything. I have to press the button again for the start menu to appear. I'm also holding the right trigger down for throttle, so maybe that has something to do with it? I don't know.

**Comment by ToruTheRedFox** [comment]
it seems to be much more likely to happen when holding throttle

**Comment by ToruTheRedFox** [comment]
just tested with Carbon and same issue is present there

**Comment by HarGabt** [comment]
> more likely to happen when holding throttle

Throttle has nothing to do with it. As noticed in Aero's gif, it has to do with steering with left stick. 
I just tried to replicate this issue and it is present. Maybe it has to do with different events that try in some way to overlay one another (turning left and going left/right in pause menu), i dunno.

**Comment by HarGabt** [comment]
Tested it more and found out the reason that may be closer to truth: if you abuse any analog axis (both triggers or left stick (with right stick it doesn't work)) by constantly changing its state, this makes it impossible to pause the game. Any other event like resetting car with Back button is not affected.

**Comment by xan1242** [comment]
![image](https://user-images.githubusercontent.com/8014093/165900358-b5c9a698-5b0d-41fd-819a-fbe5c1beb982.png)

Here's an example of a poll when it happens.

For 1 tick only, the HUDACTION_PAUSEREQUEST value is set to 1.0 when the START button is pressed.

As you can see, the polling itself is fine, it's the game misbehaving and not reading it in time. A workaround will be necessary to fix this... I'm not certain of the game's behavior on this but it may just react to changes in the value and not actually care about it being 1 or 0.

**Comment by xan1242** [comment]
Interestingly, unbinding the analog actions makes the HUD actions a lot more reactive...

So as an attempt to fix this, HUD actions will exclusively be polled before anything else

**Comment by xan1242** [comment]
The action doesn't even seem to reach the queue at all. FEngHud::JoyHandle doesn't even react in time for some reason...

**Comment by xan1242** [comment]
OK so the reason for this behavior is because of the way FEngHud::JoyHandle handles the ActionQueue.

Instead of processing each action one by one, it instead just takes 1 and pops the rest of the queue and doesn't care about the rest.

The way I made the plugin generate the input maps affects the way the queue is generated as well, hence a lot of analog actions in the queue could clog it up and make the FEngHud not react in time.

The workaround is to jump right after the ActionQueue::PopAction back to the beginning where it checks ActionQueue::IsEmpty. I hadn't yet noticed any side effects from this so I'll implement it in the next version.

**Comment by xan1242** [comment]
fixed with [commit 6ad099f](https://github.com/xan1242/NFS-XtendedInput/commit/6ad099fc3373d22868a94569f252d16e4406c11f)

Closing
